app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt

```diff
@@ -0,0 +1,418 @@
+package io.zenandroid.onlinego.ui.screens.puzzle
+
+import android.text.format.DateUtils.getRelativeTimeSpanString
+import androidx.compose.animation.ExperimentalAnimationApi
+import androidx.compose.foundation.*
+import androidx.compose.foundation.layout.*
+import androidx.compose.foundation.lazy.LazyColumn
+import androidx.compose.foundation.lazy.LazyRow
+import androidx.compose.foundation.lazy.items
+import androidx.compose.foundation.lazy.rememberLazyListState
+import androidx.compose.foundation.text.KeyboardOptions
+import androidx.compose.material.*
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.filled.*
+import androidx.compose.runtime.*
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.text.TextStyle
+import androidx.compose.ui.text.font.FontFamily
+import androidx.compose.ui.text.font.FontStyle
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.input.*
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import com.google.accompanist.pager.*
+import io.zenandroid.onlinego.R
+import io.zenandroid.onlinego.data.model.local.PuzzleCollection
+import io.zenandroid.onlinego.data.model.local.VisitedPuzzleCollection
+import io.zenandroid.onlinego.data.repositories.SettingsRepository
+import io.zenandroid.onlinego.gamelogic.RulesManager
+import io.zenandroid.onlinego.gamelogic.Util.toCoordinateSet
+import io.zenandroid.onlinego.ui.composables.Board
+import io.zenandroid.onlinego.ui.composables.RatingBar
+import io.zenandroid.onlinego.ui.screens.puzzle.PuzzleDirectoryAction.*
+import io.zenandroid.onlinego.ui.screens.puzzle.PuzzleDirectorySort.*
+import io.zenandroid.onlinego.utils.convertCountryCodeToEmojiFlag
+import org.commonmark.node.*
+import org.koin.core.context.GlobalContext
+import java.time.Instant.now
+import java.time.temporal.ChronoUnit.*
+import kotlin.reflect.KClass
+import kotlin.reflect.full.createInstance
+import kotlin.reflect.full.safeCast
+
+@Composable
+fun PuzzleDirectoryScreen(
+    state: PuzzleDirectoryState,
+    onCollection: (PuzzleCollection) -> Unit,
+    onBack: () -> Unit,
+) {
+    val listState = rememberLazyListState()
+    var filterText by remember { mutableStateOf(TextFieldValue()) }
+    var sortField by remember { mutableStateOf<PuzzleDirectorySort>(RatingSort(false)) }
+    val resultCollections = filterText.text.lowercase().let { query ->
```

Comment:
This has the potential to be run every frame, at the very least it should be remembered, but the proper way to do it is to offload this logic to the ViewModel. The Composable should be concerned only with displaying stuff, not filtering or really remembering much.

Reply:
Agreed, ok